### ISSUE TITLE
Rejection of a Promise that depends on another promise is extremely slow in V8/Chrome.

### ISSUE THREAD

**Issue by dtinth** [petkaantonov/bluebird#501]
The Story
-------------

First, I created a Promise, called `x`.

```js
var x = new Promise(function(resolve) {
  setTimeout(resolve, 1000)
})
```

Then I created a function that processes a task, but it depends on `x`, but they all fail:

```js
function y(task) {
  return x.then(function() { throw new Error('it failed') })
}
```

I have 1000 tasks, and I process them all.

```js
Promise.all(tasks.map(y)).then(...).catch(...)
```

The Problem
-----------------

At the moment `x` is resolved, on Google Chrome with a minified build, __the browser freezes for ~10 seconds__.

The Test Case
-------------

https://rawgit.com/dtinth/39520b129f143c66bd40/raw/index.html

The Weirdness
--------------------

- From my testing, __this only happens in Google Chrome__, including Chrome Canary.
    - It runs smoothly on Firefox and Safari.
- When using native promises, this problem doesn't happen.
- It __only happens in minified builds__.
    - When I use `bluebird.js` instead of `bluebird.min.js`, the problem disappears.
    - Nevertheless, when I use `bluebird` in webpack environment, the problem still appears.

The Profile
--------------

On my MacBook, it takes __9 seconds__ to settle the Promise.

![screen shot 2015-02-18 at 21 02 47](https://cloud.githubusercontent.com/assets/193136/6248379/9537e5ee-b7b1-11e4-9ff5-7231acabbbb2.png)

Upon zooming, most time is spent on `_rejectCallback`.

![screen shot 2015-02-18 at 21 04 13](https://cloud.githubusercontent.com/assets/193136/6248393/c4564c94-b7b1-11e4-906c-ce721935edba.png)

The Variations
-----------------

If `y` returns a promise that's eventually fulfilled (instead of rejected), this problem doesn't happen.

If each `y` depends on different `x`, this problem doesn't happen. What I mean is that `x` and `y` looks like this instead:

```js
function x() {
  return new Promise(function(resolve) {
    setTimeout(resolve, 1000)
  })
}

function y(task) {
  return x().then(function() { throw new Error('it failed') })
}
```

**Comment by benjamingr** [comment]
Can you reproduce this in node?

**Comment by dtinth** [comment]
Yes, this is reproducable in Node.js using the following test code:

```js
var Promise = require('bluebird')

var x = new Promise(function(resolve) {
  setTimeout(function() {
    console.log('will now resolve!')
    resolve()
  }, 1000)
})

function y(task) {
  return x.then(function() { throw new Error('it failed') })
}

var tasks = []
for (var i = 0; i < 1000; i ++) tasks.push('Task ' + i)

Promise.all(tasks.map(y))
.then(function() { console.log('done') })
.catch(function() { console.log('caught') })
```

**Comment by phpnode** [comment]
@dtinth that particular test case did not reproduce the bug for me on iojs / node 0.11.x / node 0.10.x

**Comment by dtinth** [comment]
@phpnode 

It also happens on iojs 1.0.3. I reproduced this on Mac OS X 10.10.1 with Bluebird 2.9.9.

This is also reproducible on Ubuntu 13.10 running Node v0.10.28 and Bluebird 2.9.9.

**Comment by phpnode** [comment]
@dtinth ah, sorry, you're right, I was testing on `2.4.x`, was able to reproduce with `2.9.9`

**Comment by dtinth** [comment]
By trying to install various versions, this problem started appearing since 2.8.2.

```
▹ npm install bluebird@2.8.1
bluebird@2.8.1 node_modules/bluebird

▹ time node test.js
will now resolve!
caught
node test.js  0.12s user 0.03s system 12% cpu 1.143 total

▹ npm install bluebird@2.8.2
bluebird@2.8.2 node_modules/bluebird

▹ time node test.js
will now resolve!
caught
node test.js  8.98s user 0.15s system 89% cpu 10.224 total
```

**Comment by petkaantonov** [comment]
@dtinth I will push a 2.9.10 release as soon as the browser tests come back out, no need to revert to 2.8.1 :)

**Comment by petkaantonov** [comment]
@dtinth Fixed in 2.9.10

**Comment by dtinth** [comment]
Thank you for fixing this very quickly. :)
